blog-post/blog-post.md

```diff
@@ -0,0 +1,738 @@
+```
+
+`AuctionState` has four alternative value constructors: `NotStarted` specifies that the auction has
+not yet started and stores its reserve price, `Active` denotes that the auction has begun and it
+stores the current highest bid and the address that made that bid, `Ended` represents an auction
+that has finished successfully and holds the highest bid and the address of the winner, and
+`Cancelled` represents a cancelled auction, holding the highest bid and winning address at the time
+of cancellation.
+
+Now we can define a `processAuction` function that transitions the state based on the current state
+and `msg.value`. The `match` statement lets us perform an exhaustive case analysis over each
+possible alternative state. The `_` case at the end of the match is a default that handles any
+remaining states that have not yet been explicitly matched. Exhaustiveness can be enforced by the
+compiler, ensuring that every possible state is handled exactly once.
+
+```solidity
+function processAuction(state: State) -> State {
+    match state {
+    | NotStarted(reserve) =>
+        require(msg.value >= reserve);
+        return Active(msg.value, msg.sender);
+    | Active(currentBid, bidder) =>
+        require(msg.value > currentBid);
+        transferFunds(bidder, currentBid);
+        return Active(msg.value, msg.sender);
+    | _ => return state;
+    }
+}
+```
+
+### Generics and type classes
+
+Core Solidity introduces two new mechanisms for code sharing and polymorphism: generics and
+type classes.
+
+Generics implement parametric polymorphism: they enable us to write functions and data structures
+that behave in a uniform way for all types. As an example, we can define a polymorphic `identity`
+function:
+
+```solidity
+forall T . function identity(x : T) -> T {
+    return x;
+}
+```
+
+We can also define generic types, like this this `Result` type that is parameterised by the type of
+the payload in the error case:
+
+```solidity
+data Result(T) = Ok | Err(T)
+```
+
+Generics are powerful, but by themselves quite limited. Most interesting operations are not defined
+for all types. Type classes (a.k.a Traits) are the solution: they let us define an overloaded type
+specific implementation for the same function name, and when combined with class constraints let us
+define generic functions that are polymorphic over a restricted subset of types.
+
+A type class is simply an interface specification. Consider the following definition of a class of
+types that can be multiplied:
+
+```solidity
+forall T . class T:Mul {
+    function mul(lhs : T, rhs : T) -> T;
+}
+```
+
+Instead of the concrete `wmul` function that we defined above for our `wad` fixed point type, it
+would be more idiomatic to defined an instance (a.k.a `impl`) of the `Mul` type class for `wad`
+(giving us a uniform syntax for multiplication across all types, and allowing us to use our `wad`
+type in functions that are generic over any instance of the `wad` type class):
```

Comment:
```suggestion
Instead of the concrete `wmul` function that we defined above for our `wad` fixed point type, it
is more idiomatic to define an instance (a.k.a `impl`) of the `Mul` type class for `wad`. This gives us a uniform syntax for multiplication across all types, and allows us to use our `wad`
type in functions that are generic over any instance of the `wad` type class:
```